### PR TITLE
Adding notes and external extraction (Issue 168)

### PR DESCRIPTION
**Pull request by qiyunzhu**
Tackling issue #168 
 - Add input box "notes" (done)
 - Add check box "externally extracted" (partial)

### CODE REVIEW

**Comment by qiyunzhu** [comment]
It worked in my local system. Maybe there is a trigger to update the database for Travis?

**Comment by qiyunzhu** [comment]
Checkbox added but couldn't figure out how to get this to work with db.

**Comment by tanaes** [comment]
Do you have any other clues about what you mean by this? It may be that you need to add a notes field to the test db. It should be somewhere around [this line](https://github.com/jdereus/labman/blob/6a34eefc3b6f280bb76cdb3da885a6c429d022e2/labman/db/support_files/populate_test_db.sql#L483) of `populate_test_db.sql`. I can help with that when I get back this thursday.

**Comment by qiyunzhu** [comment]
@tanaes In Amanda's pull request this `notes` field was already added to the database. That was done thru a [db patch](https://github.com/jdereus/labman/pull/221/files#diff-3ab77a809354c9caf50cd1b1753bc00eR130). But it seemed that Travis doesn't take it..

**Comment by qiyunzhu** [comment]
Checkbox updated but still not working with the database, which does not allow `None` values being added.

**Comment by tanaes** [comment]
Looks to me like the problem is that the greyed-out 'equipment id' boxes are passing `None`, and there's currently no scope for them to be `None` when creating the `GDNAExtractionProcess`:

```python
AN ERROR HAS OCCURED!
Please copy the following into an issue in our github repository
Error
qiita.equipment with ID 'None' does not exist
Traceback
Traceback (most recent call last): 
File "/Users/jonsanders/Development/anaconda/envs/labman/lib/python3.5/site-packages/tornado/web.py", line 1541, in _execute result = method(*self.path_args, **self.path_kwargs) 
File "/Users/jonsanders/Development/anaconda/envs/labman/lib/python3.5/site-packages/tornado/web.py", line 2949, in wrapper return method(self, *args, **kwargs) 
File "/Users/jonsanders/Development/git_sw/labman/labman/gui/handlers/process_handlers/gdna_extraction_process.py", line 78, in post for pid, kf, ep, ept, kit, p_name, nt in json_decode(plates_info)] 
File "/Users/jonsanders/Development/git_sw/labman/labman/gui/handlers/process_handlers/gdna_extraction_process.py", line 78, in for pid, kf, ep, ept, kit, p_name, nt in json_decode(plates_info)] 
File "/Users/jonsanders/Development/git_sw/labman/labman/db/base.py", line 40, in __init__ raise exceptions.LabmanUnknownIdError(self._table, id_) 
labman.db.exceptions.LabmanUnknownIdError: qiita.equipment with ID 'None' does not exist 
Request Info
host_name: localhost
_start_time: 1525708928.274381
protocol: http
method: POST
query_arguments: {}
_cookies: Set-Cookie: _xsrf=2|14837f86|32227ba985c9822af9064e95e0b9ff1e|1523374667 Set-Cookie: user="2|1:0|10:1525708909|4:user|20:InRlc3RAZm9vLmJhciI=|4cd048d91acd28fdf6aa7925cccb492361f5edcf3759ee1798a44581a6262aaa" Set-Cookie: username-localhost-8889="2|1:0|10:1525466644|23:username-localhost-8889|44:Y2RkOWQ1N2MyMmY1NDY5Yzg0NTBhN2RiYjc1YmViOGI=|f21aed0f04f94d13846652a7d54b7926c15ad5adb681f8ffbad870118a109933" Set-Cookie: username-localhost-8892="2|1:0|10:1525387616|23:username-localhost-8892|44:MDFmM2I1YTdhZjljNGU4YjgxMTAyYTM3MzdmMGRlZmY=|82ce1232b8cdef2e910c548a466bcced68f2e56987bc4b7bb9bd1f7c8e7bc26d"
server_connection: 
body_arguments: {'volume': [b'100'], 'extraction_date': [b'05/07/2018'], 'plates_info': [b'[["21",null,null,null,null,"Test plate 1 (gDNA)",""]]']}
connection: 
arguments: {'extraction_date': [b'05/07/2018'], 'volume': [b'100'], 'plates_info': [b'[["21",null,null,null,null,"Test plate 1 (gDNA)",""]]']}
body: b'extraction_date=05%2F07%2F2018&volume=100&plates_info=%5B%5B%2221%22%2Cnull%2Cnull%2Cnull%2Cnull%2C%22Test+plate+1+(gDNA)%22%2C%22%22%5D%5D'
host: localhost:8080
```

@ElDeveloper, @AmandaBirmingham, and thoughts on how best to resolve this? I can think of two possibilities:

1. Modify the [equipment lookup function](https://github.com/jdereus/labman/blob/a085877cb6b5e708d118ccb2e9118aad884e30de/labman/db/process.py#L453-L462) in this method to check explicitly for `None` and return an appropriate value (would that be 'Not Applicable'?)
2. Have a 'Not Applicable' value existing in the database and modify the html interface page to return that specific value when 'Externally extracted' is checked.

What do you think?

**Comment by AmandaBirmingham** [comment]
Several things:

1) It appears the actual test that is failing for this PR is https://github.com/jdereus/labman/blob/723133fcb80ed87a0721f1ec0d12204e4f9ec518/labman/gui/handlers/process_handlers/test/test_gdna_extraction_process.py#L36-L40
... because no notes value is being passed in.  Changing the definition of "data" to  

```
        data = {'extraction_date': '01/20/2018', 'volume': 10,
                'plates_info': json_encode(
                    [['21', 11, 6, 15, '157022406', 'new gdna plate',
                      'a note']])}
```
... makes this test pass, after which all tests pass (at least for me :) .

2) However, the error Jon pointed out above occurs in the interface--but isn't being surfaced by the tests.  This suggests to me that we need to add another test case for test_post_gdna_extraction_process_handler to exercise the "externally extracted"  path (where plates_info is [["21",null,null,null,null,"Test plate 1 (gDNA)",""]]).
3) As to how to actually make that externally extracted use case *work*, I strongly prefer Jon's suggestion 2 above ("Have a 'Not Applicable' value existing in the database and modify the html interface page to return that specific value when 'Externally extracted' is checked.").  I think letting the equipment be None is dangerous.

**Comment by qiyunzhu** [comment]
This should fix the issue! We have gone with suggestion 2 and added a 'None' (aka not applicable) value in the database for equipment and reagent, and the handler and interface are updated accordingly.

**Comment by qiyunzhu** [comment]
This first error is
```
ERROR: test_comment_well (labman.db.tests.test_process.TestSamplePlatingProcess)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/jdereus/labman/labman/db/tests/test_process.py", line 151, in test_comment_well
    tester.comment_well(8, 1, 'New notes')
  File "/home/travis/build/jdereus/labman/labman/db/process.py", line 290, in comment_well
    self.plate.get_well(row, col).composition.notes = comment
  File "/home/travis/build/jdereus/labman/labman/db/process.py", line 257, in plate
    return plate_module.Plate(plate_id)
  File "/home/travis/build/jdereus/labman/labman/db/base.py", line 40, in __init__
    raise exceptions.LabmanUnknownIdError(self._table, id_)
labman.db.exceptions.LabmanUnknownIdError: qiita.plate with ID 'None' does not exist
```
@ElDeveloper @tanaes

**Comment by ElDeveloper** [comment]
@qiyunzhu any chance you can run the tests locally and fix the failures? It seems like the errors are happening because of the new "none" types. I would recommend running `nosetests -v` to see all the local failures. Alternatively you could look at the failures in the travis build.

**Comment by qiyunzhu** [comment]
@ElDeveloper I was not able to run local test using `nosetests`. It says `ImportError: No module named 'mock'`. Is there anything I missed?

**Comment by AmandaBirmingham** [comment]
@qiyunzhu in a terminal window, with your labman conda environment activated, try running
conda install mock

**Comment by qiyunzhu** [comment]
@AmandaBirmingham Thanks for the suggestion! `mock` was already installed, and I can `import mock` in the Python environment.

**Comment by AmandaBirmingham** [comment]
@qiyunzhu Dang!  I got the same failure as you trying to run the unit tests locally, but for me just installing mock in the conda environment made all copacetic.

**Comment by qiyunzhu** [comment]
@AmandaBirmingham I just `conda install mock` again and it said `All requested packages already installed.` But then I figured out! What I need is `conda install nose`.  It is running now!

**Comment by ElDeveloper** [comment]
Can you try conda install nosetests? Looks like the nosetests executable might be installed somewhere else.

Yoshiki Vázquez-Baeza

**Comment by tanaes** [comment]
@ElDeveloper @AmandaBirmingham Would one of you have time to get another set of eyes on this PR? I pair-programmed this with @qiyunzhu.

File: labman/db/support_files/populate_test_db.sql [hunk 1]
```diff
@@ -1053,4 +1061,31 @@ BEGIN
     -- Update the combo index value
     UPDATE qiita.shotgun_primer_set SET current_combo_index = combo_idx;
 
+
+
+    -- Nothingness
```
Comment (tanaes) [inline labman/db/support_files/populate_test_db.sql]: @qiyunzhu maybe we could have a more descriptive comment here?

**Comment by ElDeveloper** [comment]
Thought I would give you guys a heads up, there are merge conflicts here.

File: labman/db/support_files/populate_test_db.sql [hunk 2]
```diff
@@ -372,6 +379,7 @@ BEGIN
         VALUES (ext_kit_composition_id, ext_kit_reagent_comp_type, '157022406')
         RETURNING reagent_composition_id INTO ext_kit_reagent_composition_id;
 
+
```
Comment (ElDeveloper) [inline labman/db/support_files/populate_test_db.sql]: Can you delete this line?

File: labman/db/support_files/populate_test_db.sql [hunk 3]
```diff
@@ -22,6 +22,13 @@ DECLARE
     rc_process_id_w                     BIGINT;
     reagent_comp_type                   BIGINT;
 
+    -- Variables for nothingness
```
Comment (ElDeveloper) [inline labman/db/support_files/populate_test_db.sql]: What about "Variables for externally extracted samples".

**Comment by qiyunzhu** [comment]
One more Travis issue:
```
======================================================================
FAIL: test_create (labman.db.tests.test_process.TestGDNAExtractionProcess)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/jdereus/labman/labman/db/tests/test_process.py", line 300, in test_create
    self.assertEqual(obs.notes, 'test note')
AssertionError: None != 'test note'
```
@tanaes

**Comment by tanaes** [comment]
@qiyunzhu please merge [this PR](https://github.com/qiyunzhu/labman/pull/3) to your branch, it should fix the issue. :)

**Comment by tanaes** [comment]
Thanks @AmandaBirmingham! I've updated per your suggestion and added documentation.

**Pull request merged**